source/slang/ir-sccp.cpp

```diff
@@ -0,0 +1,933 @@
+// ir-sccp.cpp
+#include "ir-sccp.h"
+
+#include "ir.h"
+#include "ir-insts.h"
+
+namespace Slang {
+
+
+// This file implements the Spare Conditional Constant Propagation (SCCP) optimization.
+//
+// We will apply the optimization over individual functions, so we will start with
+// a context struct for the state that we will share across functions:
+//
+struct SharedSCCPContext
+{
+    IRModule*       module;
+    SharedIRBuilder sharedBuilder;
+};
+//
+// Next we have a context struct that will be applied for each function (or other
+// code-bearing value) that we optimize:
+//
+struct SCCPContext
+{
+    SharedSCCPContext*      shared; // shared state across functions
+    IRGlobalValueWithCode*  code;   // the function/code we are optimizing
+
+    // The SCCP algorithm applies abstract interpretation to the code of the
+    // function using a "lattice" of values. We can think of a node on the
+    // lattice as representing a set of values that a given instruction
+    // might take on.
+    //
+    struct LatticeVal
+    {
+        // We will use three "flavors" of values on our lattice.
+        //
+        enum class Flavor
+        {
+            // The `None` flavor represent an empty set of values, meaning
+            // that we've never seen any indication that the instruction
+            // produces a (well-defined) value. This could indicate an
+            // instruction that does not appear to execute, but it could
+            // also indicate an instruction that we know invokes undefined
+            // behavior, so we can freely pick a value for it on a whim.
+            None,
+
+            // The `Constant` flavor represents an instuction that we
+            // have only ever seen produce a single, fixed value. It's
+            // `value` field will hold that constant value.
+            Constant,
+
+            // The `Any` flavor represents an instruction that might produce
+            // different values at runtime, so we go ahead and approximate
+            // this as it potentially yielding any value whatsoever. A
+            // more precise analysis could use sets or intervals of values,
+            // but for SCCP anything that could take on more than 1 value
+            // at runtime is assumed to be able to take on *any* value.
+            Any,
+        };
+
+        // The flavor of this value (`None`, `Constant`, or `Any`)
+        Flavor  flavor;
+
+        // If this is a `Constant` lattice value, then this field
+        // points to the IR instruction that defines the actual constant value.
+        // For all other flavors it should be null.
+        IRInst* value = nullptr;
+
+        // For convenience, we define `static` factory functions to
+        // produce values of each of the flavors.
+
+        static LatticeVal getNone()
```

Comment:
'get' has specific meaning in most typical code. For constructing things on the heap 'new' and 'create' are typical. For creating a value type 'make' seems a more suitable verb - to distinguish from create and to show it's not just getting a existing value.

Reply:
`make` seems reasonable for non-allocating factory functions, yeah. Somehow I still like `get` here because we aren't *making* a `None` value, but getting the one and only value, but you are probably right that the intent should be more clear.